Source/santad/EventProviders/FAAPolicyProcessor.mm

```diff
@@ -121,30 +251,244 @@
-static inline void PushBackIfNotTruncated(std::vector<FAAPolicyProcessor::PathTarget> &vec,
-                                          const es_file_t *dir, const es_string_token_t &name) {
-  if (!dir->path_truncated) {
-    vec.push_back({Path(dir) + "/" + Path(name), false, std::nullopt});
+FileAccessPolicyDecision FAAPolicyProcessor::ProcessTargetAndPolicy(
+    const Message &msg, const PathTarget &target,
+    const std::optional<std::shared_ptr<WatchItemPolicyBase>> optional_policy,
+    CheckIfPolicyMatchesBlock check_if_policy_matches_block,
+    SNTFileAccessDeniedBlock file_access_denied_block,
+    SNTOverrideFileAccessAction override_action) {
+  FileAccessPolicyDecision decision = ApplyOverrideToDecision(
+      ApplyPolicy(msg, target, optional_policy, check_if_policy_matches_block), override_action);
+
+  // Note: If ShouldLogDecision, it shouldn't be possible for optionalPolicy
+  // to not have a value. Performing the check just in case to prevent a crash.
+  if (ShouldLogDecision(decision) && optional_policy.has_value()) {
+    std::shared_ptr<WatchItemPolicyBase> policy = *optional_policy;
+
+    // TODO: Rate limiting
+    LogTelemetry(*policy, target, msg, decision);
+
+    if (ShouldNotifyUserDecision(decision) &&
+        (ShouldShowUI(policy) || ShouldMessageTTY(policy, msg))) {
+      SNTCachedDecision *cd = GetCachedDecision(msg->process->executable->stat);
+      SNTFileAccessEvent *event = [[SNTFileAccessEvent alloc] init];
+
+      event.accessedPath = StringToNSString(target.path);
+      event.ruleVersion = StringToNSString(policy->version);
+      event.ruleName = StringToNSString(policy->name);
+      event.fileSHA256 = cd.sha256 ?: @"<unknown sha>";
+      event.filePath = StringToNSString(msg->process->executable->path.data);
+      event.teamID = cd.teamID ?: @"<unknown team id>";
+      event.signingID = cd.signingID ?: @"<unknown signing id>";
+      event.cdhash = cd.cdhash ?: @"<unknown CDHash>";
+      event.pid = @(audit_token_to_pid(msg->process->audit_token));
+      event.ppid = @(audit_token_to_pid(msg->process->parent_audit_token));
+      event.parentName = StringToNSString(msg.ParentProcessName());
+      event.signingChain = cd.certChain;
+
+      struct passwd *user = getpwuid(audit_token_to_ruid(msg->process->audit_token));
+      if (user) event.executingUser = @(user->pw_name);
+
+      URLTextPair link_info;
+      if (generate_event_detail_link_block_) {
+        link_info = generate_event_detail_link_block_(policy);
+      }
+
+      if (ShouldShowUI(policy)) {
+        file_access_denied_block(event, OptionalStringToNSString(policy->custom_message),
+                                 link_info.first, link_info.second);
+      }
+
+      // TODO: TTY message cache
+      if (ShouldMessageTTY(policy, msg)) {
+        LogTTY(event, link_info, msg, *policy);
+      }
+    }
   }
+
+  return decision;
+}
+
+es_auth_result_t FAAPolicyProcessor::ProcessMessage(
```

Comment:
Mostly from: https://github.com/northpolesec/santa/blob/1406a927f78561233d8f5a7d93c83b6a5986978a/Source/santad/EventProviders/SNTEndpointSecurityFileAccessAuthorizer.mm#L554

Note: The target/policy pairs are now passed in instead of directly computed. I also have left ES cachability as a TODO and don't yet bubble up that logic.